src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -2171,134 +1942,150 @@ tx_gateway_frontend::do_abort_tm_tx(
-            co_return tx::errc::coordinator_not_available;
-        } else {
-            if (changed_tx.error() == tm_stm::op_status::not_leader) {
-                co_return tx::errc::not_coordinator;
-            }
-            if (changed_tx.error() == tm_stm::op_status::timeout) {
-                co_return tx::errc::timeout;
-            }
-            co_return tx::errc::unknown_server_error;
+    if (tx.status == tx_status::ongoing || tx.status == tx_status::empty) {
+        auto update_result = co_await stm->update_transaction_status(
+          expected_term, tx.id, tx_status::preparing_abort);
+        if (update_result.has_error()) {
+            co_return map_state_update_outcome(update_result.error());
         }
     }
-    tx = changed_tx.value();
 
-    co_return co_await reabort_tm_tx(stm, expected_term, tx, timeout);
+    auto abort_result = co_await abort_data(stm, expected_term, tx, timeout);
+    if (abort_result.has_error()) {
+        co_return abort_result.error();
+    }
+
+    auto update_result = co_await stm->finish_transaction(
+      expected_term, tx.id, tx_status::completed_abort);
+    if (update_result.has_error()) {
+        co_return map_state_update_outcome(update_result.error());
+    }
+    co_return std::move(update_result.value());
 }
 
-ss::future<checked<cluster::tx_metadata, tx::errc>>
+ss::future<tx_gateway_frontend::op_result_t>
 tx_gateway_frontend::do_commit_tm_tx(
   model::term_id expected_term,
   ss::shared_ptr<cluster::tm_stm> stm,
   cluster::tx_metadata tx,
   model::timeout_clock::duration timeout,
   ss::lw_shared_ptr<available_promise<tx::errc>> outcome) {
-    if (!stm->is_actual_term(expected_term)) {
-        outcome->set_value(tx::errc::not_coordinator);
-        co_return tx::errc::not_coordinator;
-    }
+    try {
+        if (!stm->is_actual_term(expected_term)) {
+            outcome->set_value(tx::errc::not_coordinator);
+            co_return tx::errc::not_coordinator;
+        }
 
-    if (tx.status == tx_status::prepared) {
+        if (tx.status == tx_status::ongoing || tx.status == tx_status::empty) {
+            auto update_result = co_await stm->update_transaction_status(
+              expected_term, tx.id, tx_status::preparing_commit);
+            if (!update_result) {
+                auto err = map_state_update_outcome(update_result.error());
+                outcome->set_value(err);
+                co_return err;
+            }
+        }
+        // release control to the client before committing transaction on data
+        // partitions & groups as the outcome has already been decided.
         outcome->set_value(tx::errc::none);
-        co_return co_await recommit_tm_tx(stm, expected_term, tx, timeout);
+    } catch (...) {
+        vlog(
+          txlog.warn,
+          "[tx_id={}] error committing transaction - {}",
+          tx.id,
+          std::current_exception());
+        outcome->set_value(tx::errc::unknown_server_error);
```

Comment:
I was wondering if we should set something less severe here, from my reading of code an exceptional future here is only possible in cases  like runtime errors (like ooms) and such and hence the transaction state is not ambiguous and retriable.

Reply:
I see this is resolved, but the code didn't change?